client/consensus/manual-seal/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+
+use consensus_common::{
+	self, BlockImport, Environment, Proposer, BlockCheckParams,
+	ForkChoiceStrategy, BlockImportParams, BlockOrigin,
+	ImportResult, SelectChain,
+};
+use consensus_common::import_queue::{BasicQueue, CacheKeyId, Verifier, BoxBlockImport};
+use sr_primitives::traits::{Block as BlockT};
+use sr_primitives::Justification;
+use parking_lot::Mutex;
+use futures::prelude::*;
+use transaction_pool::txpool::{self, Pool as TransactionPool};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub mod rpc;
+use rpc::EngineCommand;
+
+/// The synchronous block-import worker of the engine.
+pub struct ManualSealBlockImport<I> {
+	inner: I,
+}
+
+impl<I> From<I> for ManualSealBlockImport<I> {
+	fn from(i: I) -> Self {
+		ManualSealBlockImport { inner: i }
+	}
+}
+
+impl<B: BlockT, I: BlockImport<B>> BlockImport<B> for ManualSealBlockImport<I> {
+	type Error = I::Error;
+
+	fn check_block(&mut self, block: BlockCheckParams<B>) -> Result<ImportResult, Self::Error>
+	{
+		self.inner.check_block(block)
+	}
+
+	fn import_block(
+		&mut self,
+		block: BlockImportParams<B>,
+		cache: HashMap<CacheKeyId, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		// TODO: strip out post-digest.
+
+		self.inner.import_block(block, cache)
+	}
+}
+
+/// The verifier for the manual seal engine; instantly finalizes.
+struct ManualSealVerifier;
+
+impl<B: BlockT> Verifier<B> for ManualSealVerifier {
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let import_params = BlockImportParams {
+			origin,
+			header,
+			justification,
+			post_digests: Vec::new(),
+			body,
+			finalized: true,
+			auxiliary: Vec::new(),
+			fork_choice: ForkChoiceStrategy::LongestChain,
+			allow_missing_state: false,
+		};
+
+		Ok((import_params, None))
+	}
+}
+
+/// Instantiate the import queue for the manual seal consensus engine.
+pub fn import_queue<B: BlockT>(block_import: BoxBlockImport<B>) -> BasicQueue<B>
+{
+	BasicQueue::new(
+		ManualSealVerifier,
+		block_import,
+		None,
+		None,
+	)
+}
+
+/// Creates the background authorship task for the manual seal engine.
+pub async fn run_manual_seal<B, E, A, C, S>(
+	block_import: BoxBlockImport<B>,
+	env: E,
+	pool: Arc<TransactionPool<A>>,
+	seal_block_channel: S,
+	select_chain: C,
+	inherent_data_providers: inherents::InherentDataProviders,
+)
+	where
+		B: BlockT + 'static,
+		E: Environment<B> + 'static,
+		A: txpool::ChainApi + 'static,
+		S: Stream<Item=EngineCommand> + 'static,
+		C: SelectChain<B> + 'static,
+{
+	let block_import = Arc::new(Mutex::new(block_import));
+	let env = Arc::new(Mutex::new(env));
+	let select_chain = Arc::new(select_chain);
+	let inherent_data_providers = Arc::new(inherent_data_providers);
+	let moved_pool = pool.clone();
+
+	seal_block_channel
```

Comment:
Can't this just be a loop?

Reply:
The RPC sends the block authorship task commands over a `futures::mpsc`, also with the way the api is designed, I've implemented an instant_seal authorship task.


https://github.com/paritytech/substrate/pull/4143/files#diff-4cecb2a4202c6cedf89a9417017c3ac0R215-R250

Reply:
I mean instead of `for_each` there should be a way to write a loop that consumes elements of the stream using `async/await`, no?

Reply:
I can look into this